src/Pretty.hs

```diff
@@ -663,15 +677,157 @@ instance Pretty VarInfo Scope where
 instance Pretty VarInfo ScopeSet where
   pp env scs =
     let (allPhases, phases) = contents scs
-    in text "⟨" <> align (group (ppSet allPhases <> text "," <> line <> ppMap (ppSet <$> phases) <> "⟩"))
+    in text "⟨" <> align ((ppSet allPhases <> text "," <> line <> ppMap (ppSet <$> phases) <> "⟩"))
 
     where
-      commaSep = group . concatWith (\x y -> x <> text "," <> line <> y)
+      commaSep = concatWith (\x y -> x <> text "," <> line <> y)
       ppSet s =
         text "{" <> commaSep (map (pp env) (Set.toList s)) <> text "}"
       ppMap m =
-        group (vsep [group (viaShow k <+> text "↦" <> line <> v) | (k, v) <- St.toList m])
+        (vsep [(viaShow k <+> text "↦" <> line <> v) | (k, v) <- St.toList m])
 
 
 instance Pretty VarInfo KlisterPathError where
   pp _ = ppKlisterPathError
+
+-- -----------------------------------------------------------------------------
+-- StackTraces
+
+instance Pretty VarInfo EState where
+  pp env st = printStack env st
+
+instance Pretty VarInfo Kont where
+  pp env k = hardline <> text "----" <+> printKont env k
+
+printStack :: Env Var () -> EState -> Doc VarInfo
+printStack e (Er err _env k) =
+  vsep [ pp e err
+       , text "stack trace:"
+       ] <> pp e k
+printStack _ Up{}   = hang 2 $ text "up"
+printStack _ Down{} = hang 2 $ text "down"
+
+printKont :: Env Var () -> Kont -> Doc VarInfo
+-- the basics
+printKont _ Halt               = text "Halt"
```

Comment:
In empty traces, this was a bit confusing. What about suppressing it, and then special-casing empty continuations at the outside?